include/seqan3/index/fm_index_iterator.hpp

```diff
@@ -0,0 +1,250 @@
+
+protected:
+    using node_type = detail::fm_index_iterator_node<index_t>;
+
+    const index_type * index;
+    size_type parent_lb, parent_rb;
+    node_type node;
+
+    size_type offset() const
+    {
+        return index->m_index.size() - depth() - 1; // since the string is reversed during construction
+    }
+
+public:
+
+    fm_index_iterator() = delete;
+    fm_index_iterator(fm_index_iterator const &) = default;
+    fm_index_iterator & operator=(fm_index_iterator const &) = default;
+    fm_index_iterator(fm_index_iterator &&) = default;
+    fm_index_iterator & operator=(fm_index_iterator &&) = default;
+
+    fm_index_iterator(index_t const & _index) : index(&_index), node({0, _index.m_index.size() - 1, 0, 0})
+    {}
+
+    // TODO: cannot compare indices yet (not supported by sdsl)
+    bool operator==(fm_index_iterator const & rhs) const
+    {
+        // parent_lb/parent_rb might be uninitialized in a root node
+        return node == rhs.node && ((parent_lb == rhs.parent_lb && parent_rb == rhs.parent_rb) || depth() == 0);
+    }
+
+    bool operator!=(fm_index_iterator const & rhs) const
+    {
+        return !(*this == rhs);
+    }
+
+    // TODO: implement first backward search more efficiently if iterator is at root node!
+    // NOTE: whenever false is returned, it is guaranteed that the iterator remained untouched
+    bool down()
+    {
+        typename index_type::comp_char_type c = 1; // NOTE: start with 0 or 1 depending on implicit_sentintel
+        typename index_type::size_type _lb, _rb;
+        for (; c < index->m_index.sigma && !sdsl::backward_search(index->m_index, node.lb, node.rb, index->m_index.comp2char[c], _lb, _rb); ++c) {}
+        if (c != index->m_index.sigma)
+        {
+            parent_lb = node.lb;
+            parent_rb = node.rb;
+            node = {_lb, _rb, node.depth + 1, c};
+            return true;
+        }
+        return false;
+    }
+
+    bool down(typename index_type::char_type const & c)
+    {
+        typename index_type::size_type _lb, _rb;
+
+        auto c_char = to_rank(c) + 1;
+        auto c_comp = index->m_index.char2comp[c_char];
+
+        // character does not occur in text / index
+        if (!c_comp) // TODO: [[unlikely]]
+            return false;
+
+        if (sdsl::backward_search(index->m_index, node.lb, node.rb, c_char, _lb, _rb))
+        {
+            parent_lb = node.lb;
+            parent_rb = node.rb;
+            node = {_lb, _rb, node.depth + 1, c_comp};
+            return true;
+        }
+        return false;
+    }
+
+    // NOTE: range cannot be empty
+    // TODO: requires clause does not work with == ... syntax error?
+    // TODO: enforce the innermost type of char_type to be convertable to char_type
+    template <range_concept pattern_t>
+        requires std::is_same_v<innermost_value_type_t<pattern_t>, typename index_t::char_type>
+        // requires (innermost_value_type_t<pattern_t> == typename index_t::char_type)
```

Comment:
My educated guess would be 
`std::is_same_v<innermost_value_type_t<pattern_t>, typename index_t::char_type>`
Not really sure, though!